From 8b86d2dee9ee0f949d5ebf196782fcb0a2225e7d Mon Sep 17 00:00:00 2001
From: Yanqiang Miao <miaoyq_2010@163.com>
Date: Thu, 13 Jan 2022 09:38:07 +0800
Subject: The output of "istioctl ps" cmd show the cluster(k8s) name which a
 pod belongs (#36292)

* add clusterID in the response of /debug/syncz

Signed-off-by: Yanqiang Miao <miaoyq_2010@163.com>

* add clusterID in output of istioctl ps

Signed-off-by: Yanqiang Miao <miaoyq_2010@163.com>

* add release note

Signed-off-by: Yanqiang Miao <miaoyq_2010@163.com>
---
 istioctl/cmd/proxystatus_test.go                         | 6 +++---
 istioctl/pkg/writer/pilot/status.go                      | 9 ++++++---
 istioctl/pkg/writer/pilot/status_test.go                 | 4 ++++
 .../pkg/writer/pilot/testdata/multiStatusMultiPilot.txt  | 8 ++++----
 .../pkg/writer/pilot/testdata/multiStatusSinglePilot.txt | 6 +++---
 istioctl/pkg/writer/pilot/testdata/singleStatus.txt      | 4 ++--
 .../pkg/writer/pilot/testdata/singleStatusFallback.txt   | 4 ++--
 pilot/pkg/xds/debug.go                                   | 2 ++
 releasenotes/notes/36290.yaml                            | 7 +++++++
 9 files changed, 33 insertions(+), 17 deletions(-)
 create mode 100644 releasenotes/notes/36290.yaml

diff --git a/istioctl/cmd/proxystatus_test.go b/istioctl/cmd/proxystatus_test.go
index e88f59e363..d4df63b8e3 100644
--- a/istioctl/cmd/proxystatus_test.go
+++ b/istioctl/cmd/proxystatus_test.go
@@ -24,11 +24,11 @@ func TestProxyStatus(t *testing.T) {
 	cases := []execTestCase{
 		{ // case 0
 			args:           strings.Split("proxy-status", " "),
-			expectedString: "NAME     CDS     LDS     EDS     RDS     ISTIOD",
+			expectedString: "NAME     CLUSTER     CDS     LDS     EDS     RDS     ISTIOD",
 		},
 		{ // case 1 short name "ps"
 			args:           strings.Split("ps", " "),
-			expectedString: "NAME     CDS     LDS     EDS     RDS     ISTIOD",
+			expectedString: "NAME     CLUSTER     CDS     LDS     EDS     RDS     ISTIOD",
 		},
 		{ // case 2: supplying nonexistent pod name should result in error with flag
 			args:          strings.Split("proxy-status deployment/random-gibberish", " "),
@@ -48,7 +48,7 @@ func TestProxyStatus(t *testing.T) {
 		},
 		{ // case 6: new --revision argument
 			args:           strings.Split("proxy-status --revision canary", " "),
-			expectedString: "NAME     CDS     LDS     EDS     RDS     ISTIOD",
+			expectedString: "NAME     CLUSTER     CDS     LDS     EDS     RDS     ISTIOD",
 		},
 		{ // case 7: supplying type that doesn't select pods should fail
 			args:          strings.Split("proxy-status serviceaccount/sleep", " "),
diff --git a/istioctl/pkg/writer/pilot/status.go b/istioctl/pkg/writer/pilot/status.go
index e2f34e315c..84cfb33385 100644
--- a/istioctl/pkg/writer/pilot/status.go
+++ b/istioctl/pkg/writer/pilot/status.go
@@ -89,7 +89,7 @@ func (s *StatusWriter) PrintSingle(statuses map[string][]byte, proxyName string)
 
 func (s *StatusWriter) setupStatusPrint(statuses map[string][]byte) (*tabwriter.Writer, []*writerStatus, error) {
 	w := new(tabwriter.Writer).Init(s.Writer, 0, 8, 5, ' ', 0)
-	_, _ = fmt.Fprintln(w, "NAME\tCDS\tLDS\tEDS\tRDS\tISTIOD\tVERSION")
+	_, _ = fmt.Fprintln(w, "NAME\tCLUSTER\tCDS\tLDS\tEDS\tRDS\tISTIOD\tVERSION")
 	fullStatus := make([]*writerStatus, 0, len(statuses))
 	for pilot, status := range statuses {
 		var ss []*writerStatus
@@ -103,6 +103,9 @@ func (s *StatusWriter) setupStatusPrint(statuses map[string][]byte) (*tabwriter.
 		fullStatus = append(fullStatus, ss...)
 	}
 	sort.Slice(fullStatus, func(i, j int) bool {
+		if fullStatus[i].ClusterID != fullStatus[j].ClusterID {
+			return fullStatus[i].ClusterID < fullStatus[j].ClusterID
+		}
 		return fullStatus[i].ProxyID < fullStatus[j].ProxyID
 	})
 	return w, fullStatus, nil
@@ -120,8 +123,8 @@ func statusPrintln(w io.Writer, status *writerStatus) error {
 		// but it is better than not providing any information.
 		version = status.ProxyVersion + "*"
 	}
-	_, _ = fmt.Fprintf(w, "%v\t%v\t%v\t%v\t%v\t%v\t%v\n",
-		status.ProxyID, clusterSynced, listenerSynced, endpointSynced, routeSynced, status.pilot, version)
+	_, _ = fmt.Fprintf(w, "%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\n",
+		status.ProxyID, status.ClusterID, clusterSynced, listenerSynced, endpointSynced, routeSynced, status.pilot, version)
 	return nil
 }
 
diff --git a/istioctl/pkg/writer/pilot/status_test.go b/istioctl/pkg/writer/pilot/status_test.go
index 51aa9b1a55..74c69d0f35 100644
--- a/istioctl/pkg/writer/pilot/status_test.go
+++ b/istioctl/pkg/writer/pilot/status_test.go
@@ -171,6 +171,7 @@ func TestStatusWriter_PrintSingle(t *testing.T) {
 func statusInput1() []xds.SyncStatus {
 	return []xds.SyncStatus{
 		{
+			ClusterID:     "cluster1",
 			ProxyID:       "proxy1",
 			IstioVersion:  "1.1",
 			ClusterSent:   preDefinedNonce,
@@ -186,6 +187,7 @@ func statusInput1() []xds.SyncStatus {
 func statusInput2() []xds.SyncStatus {
 	return []xds.SyncStatus{
 		{
+			ClusterID:     "cluster2",
 			ProxyID:       "proxy2",
 			IstioVersion:  "1.1",
 			ClusterSent:   preDefinedNonce,
@@ -203,6 +205,7 @@ func statusInput2() []xds.SyncStatus {
 func statusInput3() []xds.SyncStatus {
 	return []xds.SyncStatus{
 		{
+			ClusterID:     "cluster3",
 			ProxyID:       "proxy3",
 			IstioVersion:  "1.1",
 			ClusterSent:   preDefinedNonce,
@@ -219,6 +222,7 @@ func statusInput3() []xds.SyncStatus {
 func statusInputProxyVersion() []xds.SyncStatus {
 	return []xds.SyncStatus{
 		{
+			ClusterID:     "cluster2",
 			ProxyID:       "proxy2",
 			ProxyVersion:  "1.1",
 			ClusterSent:   preDefinedNonce,
diff --git a/istioctl/pkg/writer/pilot/testdata/multiStatusMultiPilot.txt b/istioctl/pkg/writer/pilot/testdata/multiStatusMultiPilot.txt
index c000bded3d..817b1f6957 100644
--- a/istioctl/pkg/writer/pilot/testdata/multiStatusMultiPilot.txt
+++ b/istioctl/pkg/writer/pilot/testdata/multiStatusMultiPilot.txt
@@ -1,5 +1,5 @@
-NAME       CDS                            LDS          EDS                            RDS          ISTIOD      VERSION
-proxy1     STALE                          SYNCED       SYNCED                         NOT SENT     istiod1     1.1
-proxy2     STALE                          SYNCED       STALE                          SYNCED       istiod2     1.1
-proxy3     STALE (Never Acknowledged)     NOT SENT     STALE (Never Acknowledged)     SYNCED       istiod3     1.1
+NAME       CLUSTER      CDS                            LDS          EDS                            RDS          ISTIOD      VERSION
+proxy1     cluster1     STALE                          SYNCED       SYNCED                         NOT SENT     istiod1     1.1
+proxy2     cluster2     STALE                          SYNCED       STALE                          SYNCED       istiod2     1.1
+proxy3     cluster3     STALE (Never Acknowledged)     NOT SENT     STALE (Never Acknowledged)     SYNCED       istiod3     1.1
 
diff --git a/istioctl/pkg/writer/pilot/testdata/multiStatusSinglePilot.txt b/istioctl/pkg/writer/pilot/testdata/multiStatusSinglePilot.txt
index 9d33f46edf..2abb26836a 100644
--- a/istioctl/pkg/writer/pilot/testdata/multiStatusSinglePilot.txt
+++ b/istioctl/pkg/writer/pilot/testdata/multiStatusSinglePilot.txt
@@ -1,3 +1,3 @@
-NAME       CDS       LDS        EDS        RDS          ISTIOD      VERSION
-proxy1     STALE     SYNCED     SYNCED     NOT SENT     istiod1     1.1
-proxy2     STALE     SYNCED     STALE      SYNCED       istiod1     1.1
+NAME       CLUSTER      CDS       LDS        EDS        RDS          ISTIOD      VERSION
+proxy1     cluster1     STALE     SYNCED     SYNCED     NOT SENT     istiod1     1.1
+proxy2     cluster2     STALE     SYNCED     STALE      SYNCED       istiod1     1.1
diff --git a/istioctl/pkg/writer/pilot/testdata/singleStatus.txt b/istioctl/pkg/writer/pilot/testdata/singleStatus.txt
index bb0a158001..feebb1bbbd 100644
--- a/istioctl/pkg/writer/pilot/testdata/singleStatus.txt
+++ b/istioctl/pkg/writer/pilot/testdata/singleStatus.txt
@@ -1,2 +1,2 @@
-NAME       CDS       LDS        EDS       RDS        ISTIOD      VERSION
-proxy2     STALE     SYNCED     STALE     SYNCED     istiod2     1.1
+NAME       CLUSTER      CDS       LDS        EDS       RDS        ISTIOD      VERSION
+proxy2     cluster2     STALE     SYNCED     STALE     SYNCED     istiod2     1.1
diff --git a/istioctl/pkg/writer/pilot/testdata/singleStatusFallback.txt b/istioctl/pkg/writer/pilot/testdata/singleStatusFallback.txt
index 502b76ea9c..589cce599d 100644
--- a/istioctl/pkg/writer/pilot/testdata/singleStatusFallback.txt
+++ b/istioctl/pkg/writer/pilot/testdata/singleStatusFallback.txt
@@ -1,2 +1,2 @@
-NAME       CDS       LDS        EDS       RDS        ISTIOD      VERSION
-proxy2     STALE     SYNCED     STALE     SYNCED     istiod2     1.1*
+NAME       CLUSTER      CDS       LDS        EDS       RDS        ISTIOD      VERSION
+proxy2     cluster2     STALE     SYNCED     STALE     SYNCED     istiod2     1.1*
diff --git a/pilot/pkg/xds/debug.go b/pilot/pkg/xds/debug.go
index 23e5fd73ee..48f71d5694 100644
--- a/pilot/pkg/xds/debug.go
+++ b/pilot/pkg/xds/debug.go
@@ -106,6 +106,7 @@ type AdsClients struct {
 
 // SyncStatus is the synchronization status between Pilot and a given Envoy
 type SyncStatus struct {
+	ClusterID     string `json:"cluster_id,omitempty"`
 	ProxyID       string `json:"proxy,omitempty"`
 	ProxyVersion  string `json:"proxy_version,omitempty"`
 	IstioVersion  string `json:"istio_version,omitempty"`
@@ -265,6 +266,7 @@ func (s *DiscoveryServer) Syncz(w http.ResponseWriter, _ *http.Request) {
 		if node != nil {
 			syncz = append(syncz, SyncStatus{
 				ProxyID:       node.ID,
+				ClusterID:     node.Metadata.ClusterID.String(),
 				IstioVersion:  node.Metadata.IstioVersion,
 				ClusterSent:   con.NonceSent(v3.ClusterType),
 				ClusterAcked:  con.NonceAcked(v3.ClusterType),
diff --git a/releasenotes/notes/36290.yaml b/releasenotes/notes/36290.yaml
new file mode 100644
index 0000000000..8bee10f5a1
--- /dev/null
+++ b/releasenotes/notes/36290.yaml
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+
+releaseNotes:
+- |
+  **Added** `CLUSTER` column in the output of `istioctl ps` command.
-- 
2.35.3

